livekit-agents/livekit/agents/pipeline/pipeline_agent.py

```diff
@@ -276,6 +284,9 @@ def __init__(
         self._speech_q: list[SpeechHandle] = []
         self._speech_q_changed = asyncio.Event()
 
+        self._fnc_q: list[FunctionCallsHandle] = []
+        self._fnc_q_changed = asyncio.Event()
+
         self._update_state_task: asyncio.Task | None = None
 
```

Comment:
I'm not sure if we should use a different queue.
I was more thinking about nesting all speeches together. 

Imagine this scenario:
1. User speaks and request a fnc call
2. The agent generates an answer containing this specific call
3. The fnc call contains multiple calls to the say method.
4. If the user speaks it may interrupt this specific "say" speech but not the "parent" 

With the current approach, there is also no straightforward way to map the function calls to a speech/answer?

Reply:
The advantages with nested calls is that we can make it recursive:

e.g:
Function call is called but the say method takes an `LLMStream` in input that may call other functions.
In this case we have to wait for this whole recursive process before ending the root speech

Reply:
I got two questions:

How do you think the nested_speeches should work, like a speech triggers a function call, then we add the speeches from this function call (including `say` or any nested function calls) to the `nested_speeches`?

Where should these nested_speeches be handled, a separate task inside the `_play_speech` or reusing the `_play_speech` somehow?

Reply:
> Imagine this scenario:
> User speaks and request a fnc call
The agent generates an answer containing this specific call
The fnc call contains multiple calls to the say method.
If the user speaks it may interrupt this specific "say" speech but not the "parent"

From here https://github.com/livekit/agents/blob/main/livekit-agents/livekit/agents/pipeline/pipeline_agent.py#L820 the `interrupted` flag is not used to exit the nested function call loop, so in the existing version the nested function calls are already not interruptable from my understanding. Is that correct?

Reply:
> The advantages with nested calls is that we can make it recursive

I agree with this one. If I understand correctly, what you mean is that the function calls triggered by `say` should be handled before the parent function calls while the current implementation cannot do that. I think this makes sense.

Reply:
> How do you think the nested_speeches should work, like a speech triggers a function call, then we add the speeches from this function call (including `say` or any nested function calls) to the `nested_speeches`?

Yep exactly

> Where should these nested_speeches be handled, a separate task inside the `_play_speech` or reusing the `_play_speech` somehow?

Reusing the _play_speech would be the best, so we have the same code.



I'm imagining this scenario for interruption involving nested speeches.


```
    @llm.ai_callable()
    async def get_weather(
        self,
        location: Annotated[
            str, llm.TypeInfo(description="The location to get the weather for")
        ],
    ):
          speech_handle = await agent.say("This is a random speech")
          if speech_handle.interrupted: # user can make this decision for e.g
               return
```
Don't implement this feature on this PR since speech handles are currently private but this is just to give you an order of idea that interruptions should be "per" nested speeches.

Reply:
@theomonnom I created a new PR based on your suggestion: https://github.com/livekit/agents/pull/1130